### PR TITLE
Normalize, LinearTransformation are scriptable

### LINKED ISSUE #2641: Torch script support for Compose, RandomApply and others
## 🚀 Feature

Idea is to make sure that a compose of transforms could be scripted. For example, the following should work
```python
transforms = T.Compose([
    T.RandomResizedCrop(...),
    T.Random.RandomHorizontalFlip(),
    T.Normalize(...)
])

transforms = torch.jit.scrtipt(transforms)

img = torch.Tensor(...)
out_img = transforms(img)
```

Transforms to support:
- [ ] Normalize
- [ ] Compose
- [ ] RandomApply

cc @vfdev-5

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:

- Normalize, LinearTransformation are scriptable, support batch tensors and cuda tensors
- removed `object` from class definition
- updated docs

--- 

_OLD_
 
Description:
- Compose, RandomApply, Normalize can be jit scripted
- all transforms are now derived from `torch.nn.Module`

- Checked on 
  - [x] references/classification
  - [x] references/similarity
  - [x] pytorch/examples
     - dcgan
     - fast neural style
     - super-resolution
     - imagenet
  - [x] [swav](https://github.com/facebookresearch/swav)

TODO:
- [ ] Update docs and put a note about the change and possible problems if encountered...

We need an extensive checking on possible data aug pipelines to stress the code. Any help with that and reporting bugs here is appreciated.

cc @pmeier @fmassa

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -47,13 +51,30 @@ class Compose(object):
...
+            # if something fails we keep previous behaviour
+            new_transforms = []
+            for t in transforms:
+                if callable(t) and not isinstance(t, torch.nn.Module):
+                    t = Lambda(t)
+                new_transforms.append(t)
+
+            self.transforms = torch.nn.ModuleList(new_transforms)
+        except TypeError:
+            self.transforms = transforms
```
Comment (pmeier) [inline torchvision/transforms/transforms.py:68]: I think you are a little too paranoid here. The only time we could hit a `TypeError` is if `t` is not a callable. It is true, that this raised no error before, but the resulting `Compose` would not have been valid: if you call it in such case L72 would raise the error during the forward pass. Thus, anyone that is affected by now raising the `TypeError` at instantiation wasn't using the `Compose` anyway.

```suggestion
        new_transforms = []
        for t in transforms:
            if not callable(t):
                raise TypeError
            if not isinstance(t, torch.nn.Module):
                t = Lambda(t)
            new_transforms.append(t)

        self.transforms = torch.nn.ModuleList(new_transforms)
```

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -47,13 +51,30 @@ class Compose(object):
...
+            # Try to wrap the transforms that are not nn.Module into nn.Lambda
+            # register all transforms as nn.ModuleList <-> all transforms should be nn.Module
+            # if something fails we keep previous behaviour
+            new_transforms = []
+            for t in transforms:
+                if callable(t) and not isinstance(t, torch.nn.Module):
+                    t = Lambda(t)
+                new_transforms.append(t)
+
+            self.transforms = torch.nn.ModuleList(new_transforms)
```
Comment (pmeier) [inline torchvision/transforms/transforms.py:66]: What is the motivation for using a `ModuleList` here? Given that `_forward_impl` simply passes the input sequentially through all transforms, would a `Sequential` a better fit?

**Review by pmeier — COMMENTED** [review]
I have no other pipeline to test this against. Thus, my comments are only based on this PR.

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -47,13 +51,30 @@ class Compose(object):
...
 
-    def __call__(self, img):
+        try:
+            # Try to wrap the transforms that are not nn.Module into nn.Lambda
+            # register all transforms as nn.ModuleList <-> all transforms should be nn.Module
+            # if something fails we keep previous behaviour
+            new_transforms = []
+            for t in transforms:
+                if callable(t) and not isinstance(t, torch.nn.Module):
+                    t = Lambda(t)
```
Comment (vfdev-5) [inline torchvision/transforms/transforms.py:63]: An issue here can be the following. For example, user would like to configure during the training some options of its custom transform and if it is wrapped by Lambda then previous code wont work...
```python

transforms = Compose([
    ...
    CustomTransform(some_config),  # at index 5
    ...
])

# assert hasattr(CustomTransform, "update_config")
# assert isinstance(transforms[5], Lambda)

for epoch in range(100):
     
    if epoch == 25:
        transforms[5].update_config(new_config)

```

As suggested by Philip, that we can "either enables attribute access to the underlying object or at least gives a meaningful error message".

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:66]: Here is another point for using `ModuleList` instead of `Sequential`. Previously, `Compose.transforms` is of the same type as input transforms: tuple, list. User can create transforms as a compose and then append/extend etc its content. For example, https://github.com/facebookresearch/swav/blob/master/src/multicropdataset.py#L47

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -370,36 +395,57 @@ def __repr__(self):
             format(self.padding, self.fill, self.padding_mode)
 
 
-class Lambda(object):
+class Lambda(torch.nn.Module):
     """Apply a user-defined lambda as a transform.
 
+    .. Note::
+        This class exposes ``lambd`` attributes as its own attributes:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py:402]: Question: do we want to mention it in the doc? Does this make it less of an implementation detail?

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -370,36 +395,57 @@ def __repr__(self):
...
+    def forward(self, img: Tensor) -> Tensor:
         return self.lambd(img)
 
     def __repr__(self):
         return self.__class__.__name__ + '()'
 
+    def __getattr__(self, attr):
+        if hasattr(self.lambd, attr):
+            return getattr(self.lambd, attr)
+        return super().__getattr__(attr)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py:427]: Hum... can you explain why we need to override `__getattr__`   and `__setattr__` here?

File: torchvision/transforms/transforms.py [hunk 6]
```diff
@@ -420,15 +466,13 @@ class RandomApply(RandomTransforms):
...
 
-    def __call__(self, img):
-        if self.p < random.random():
+    def forward(self, img: Tensor) -> Tensor:
+        if self.p < torch.rand(1):
             return img
-        for t in self.transforms:
-            img = t(img)
-        return img
+        return self._forward_impl(img)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py:475]: Now I see why you implemented `Compose` with `_forward_impl`. I wonder if this is really necessary, given that we only save two lines but it adds additional complexity to the user to read the function?

File: test/test_transforms.py [hunk 7]
```diff
@@ -1685,6 +1685,67 @@ def test_random_erasing(self):
...
+            return Image.blend(img, img, alpha=0.5)
+
+        t = transforms.Compose([
+            transforms.RandomResizedCrop(15),
+            transforms.Lambda(lambda x: x),
+            custom_transform,
+        ])
+
+        t.transforms.append(transforms.ToTensor())
+        t.transforms.append(lambda x: x + 2)  # THIS DOES NOT WORK
```
Comment (fmassa) [inline test/test_transforms.py:1704]: Can you explain what this doesn't work?

File: test/test_transforms.py [hunk 8]
```diff
@@ -1685,6 +1685,67 @@ def test_random_erasing(self):
...
+
+            def __call__(self, x):
+                return x + self.s
+
+        t = MyTransform(10)
+        t_lambda = transforms.Lambda(t)
+        self.assertEqual(t_lambda.s, t.s)
+        t_lambda.s = 5
+        self.assertEqual(t_lambda.s, t.s)
+        self.assertEqual(t_lambda(10), t(10))
```
Comment (fmassa) [inline test/test_transforms.py:1747]: Hum, now I think I see why you override `__getattr__`. This wasn't the case before, I suppose this is to try to keep BC with Compose, if the user wants to access those fields after we wrapped they non-nn module in a `Lambda`?

My first thought would be to avoid this type of magic as it could be confusing to the user, but I'm not sure I see a better solution for this yet, apart from encouraging all new use-cases to use `nn.Sequential`, and letting `Compose` be not recommended module to use.

Thoughts?

Comment (vfdev-5) [reply test/test_transforms.py:1747]: Yes, the reason to override `__getattr__` and `__setattr__` is to be BC with Compose. More details [above](https://github.com/pytorch/vision/pull/2645#discussion_r484452454).

File: test/test_transforms.py (hunk 7 above)
Comment (vfdev-5) [reply test/test_transforms.py:1704]: This fails due to the fact that `lambda x: x + 2` is not `nn.Module`. Here we modify constructed Compose with custom ops that are not wrapped by Lambda.
See also the discussion [above](https://github.com/pytorch/vision/pull/2645#discussion_r487965999).

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:402]: Maybe we can rephrase to say that custom ops wrapped by Lambda can behave similarly to original custom op and we do not have to use `op.lambd` to access attributes.

File: torchvision/transforms/transforms.py [hunk 9]
```diff
@@ -443,7 +467,7 @@ def __repr__(self):
 class RandomOrder(RandomTransforms):
     """Apply a list of transformations in a random order
     """
-    def __call__(self, img):
+    def forward(self, img):
         order = list(range(len(self.transforms)))
         random.shuffle(order)
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:492]: RandomTransforms becomes `nn.Module`, shouldn't we use forward instead of `__call__` ?

File: test/test_transforms_tensor.py [hunk 10]
```diff
@@ -343,6 +343,74 @@ def test_to_grayscale(self):
...
+        tensor = tensor.to(dtype=torch.float32) / 255.0
+
+        transforms_list = [
+            T.RandomApply([
+                T.RandomHorizontalFlip(),
+                T.RandomVerticalFlip(),
+            ], p=0.3),
+            T.Compose([
+                T.RandomResizedCrop(15),
+                T.RandomApply([
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_transforms_tensor.py:392]: Well, I can add a basic test for that as `RandomOrder` is not scriptable

File: torchvision/transforms/transforms.py (hunk 6 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:475]: Personnally, I'd prefer to refactor... If you think that it is better for transparency reasons to put explicit code, I'm fine with that too.

File: torchvision/transforms/transforms.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/transforms/transforms.py:492]: Yes, we should. I was just pointing out that this is not torchscriptable

**Comment by vfdev-5** [comment]
According to the discussion with @fmassa : 
- Let's keep Compose as is => no BC-break
- Update its doc to say about using `nn.Sequential` for `torch.jit.script`
- Remove `__get/setattr__` from Lambda

**Review by fmassa — APPROVED** [review]
Thanks a lot!
